osd: exclude down OSDs from main PDB when cluster is clean

elias-dbx opened the pull request:
Currently, when there are down OSDs but all PGs are clean we add the number of downed OSDs ot the main PDB maxUnavailable. For example if OSDs 1,3,5 are down and all PGs are clean we create the following PDB:
```yaml
  maxUnavailable: 4
  selector:
    matchLabels:
      app: rook-ceph-osd
```

This is potentially dangerous as there can be race conditions which can break the cluster, for example:
1. A new node is added to the cluster with 4 OSDs
2. Those OSDs are down as they have not started up and joined the cluster yet
3. All PGs are healthy as the new OSDs haven't started yet and they are new
4. The rook operator sets maxUnavailable to 5 on the OSD PDB
5. The new OSDs move from down to up as they start
6. Something in the cluster starts a disruption (eviction request)
7. Any 5 OSDs can be disrupted as maxUnavilable is still 5 and the downed OSDs are now up
8. If the 5 disrupted OSDs contain all the replicas of a PG it will cause the PG to go down

To better protect from this scenario, we can instead exclude the downed OSDs from the main PDB instead of adding to maxUnavailable.
```yaml
  maxUnavailable: 1
  selector:
    matchExpressions:
    - key: app
      operator: In
      values: ["rook-ceph-osd"]
    - key: osd
      operator: NotIn
      values: ["1", "3", "5"]
```

This allows drains to still process when there are downed OSDs and PGs are clean without allowing the non-downed OSDs to be disrupted.

<!-- Thank you for contributing to Rook! -->

<!-- STEPS TO FOLLOW:
  1. Add a description of the changes (frequently the same as the commit description)
  2. Enter the issue number next to "Resolves #" below (if there is no tracking issue resolved, **remove that section**)
  3. Review our Contributing documentation at https://rook.io/docs/rook/latest/Contributing/development-flow/
  4. Follow the steps in the checklist below, starting with the **Commit Message Formatting**.
-->

<!-- Uncomment this section with the issue number if an issue is being resolved
**Issue resolved by this Pull Request:**
Resolves #
--->

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

sp98 commented:
> Something in the cluster starts a disruption (eviction request)

I don't think this `something` can be user generated initiated activity as we don't expect them the do such disruptions along with adding new nodes! But we surely can't control what users can do. 
If the `something` is not a user initiated activity, like a node lose, then PDBs won't be helpful in the case as they only work with the eviction API (triggered with node drain).

sp98 commented:
> > I don't think this something can be user generated initiated activity as we don't expect them the do such disruptions along with adding new nodes! But we surely can't control what users can do.
> > If the something is not a user initiated activity, like a node lose, then PDBs won't be helpful in the case as they only work with the eviction API (triggered with node drain).
> 
> A simple example is the user may have installed in their cluster the aws-node-termination-handler which issues drain requests against nodes based upon scheduled events in EC2: https://github.com/aws/aws-node-termination-handler. That process is 100% automated and outside of the user's control as the drain requests are initiated from AWS generated node events.
> 
> Another example is the cluster may have multiple operators, so one operator may be adding an OSD while the another is performing a node upgrade and evicts another OSD.


> Is there a problem with the PR? I believe this is a real problem as I have personally been affected by it. Is there something you do not like with the solution I have come up with?

Hi @elias-dbx We had discussed about this `matchexpressions` solution that you have implemented. And we think its a better solution than what we currently have.  

How does the overall testing of this approach looks like?  Is the testing complete and just the PR review is left?

elias-dbx commented:
@sp98 is there a guide for testing PDB in minikube and/or writing integration tests? Maybe we could write some simple integration tests for PDB?

It looks like the issue you found is due to object validation in the k8s API server which is unfortunately not done with the fake k8s client.

sp98 commented:
> @sp98 is there a guide for testing PDB in minikube and/or writing integration tests? Maybe we could write some simple integration tests for PDB?
> 
/test/scripts/create-dev-cluster.sh might be useful for testing it in minikube.

elias-dbx commented:
Ok, from the logs I know what the issue is. But I will try and spin up a minikube so I can be sure things work OK.

sp98 commented:
@elias-dbx Did you get a chance to test this out?

elias-dbx commented:
I got minikube setup in a dev environment, so I will try out testing the changes.

elias-dbx commented:
Hey I got minikube set up in a development environment in EC2 but it is very slow as KVM isn't enabled on non-metal EC2 instances and qemu without KVM is VERY slow, like it takes 10 minutes to start a pod and unfortunately the rook cluster is unstable due to the slowness.

I'm going to try and get this working on a OSX environment and use hyperkit to see if it functions better.

What dev env do you all use? Bare metal Linux? OSX?

elias-dbx commented:
@sp98 @travisn I have updated the PR and tested using minikube. I also updated the unit tests to be more in depth to test the PDB spec in all scenarios (instead of just the default PDB).

Using minikube, this is the default PDB when an OSD is down but all PGs are clean:
```
kubectl get pdb -n rook-ceph -o yaml
apiVersion: v1
items:
- apiVersion: policy/v1
  kind: PodDisruptionBudget
  metadata:
    creationTimestamp: "2025-09-08T22:37:47Z"
    generation: 4
    name: rook-ceph-osd
    namespace: rook-ceph
    ownerReferences:
    - apiVersion: ceph.rook.io/v1
      blockOwnerDeletion: true
      controller: true
      kind: CephCluster
      name: my-cluster
      uid: d297fcb1-f6b0-4188-b888-7c5355eac36a
    resourceVersion: "3464"
    uid: 43444359-d5b4-4f32-9de1-a7a6d01ac178
  spec:
    maxUnavailable: 1
    selector:
      matchExpressions:
      - key: app
        operator: In
        values:
        - rook-ceph-osd
      - key: osd
        operator: NotIn
        values:
        - "5"
  status:
    conditions:
    - lastTransitionTime: "2025-09-08T22:38:12Z"
      message: ""
      observedGeneration: 4
      reason: SufficientPods
      status: "True"
      type: DisruptionAllowed
    currentHealthy: 5
    desiredHealthy: 4
    disruptionsAllowed: 1
    expectedPods: 5
    observedGeneration: 4
kind: List
metadata:
  resourceVersion: ""
```

And then this is the default PDB once the downed OSD comes back:
```
kubectl -n rook-ceph get pdb -o yaml
apiVersion: v1
items:
- apiVersion: policy/v1
  kind: PodDisruptionBudget
  metadata:
    creationTimestamp: "2025-09-08T22:37:47Z"
    generation: 5
    name: rook-ceph-osd
    namespace: rook-ceph
    ownerReferences:
    - apiVersion: ceph.rook.io/v1
      blockOwnerDeletion: true
      controller: true
      kind: CephCluster
      name: my-cluster
      uid: d297fcb1-f6b0-4188-b888-7c5355eac36a
    resourceVersion: "4557"
    uid: 43444359-d5b4-4f32-9de1-a7a6d01ac178
  spec:
    maxUnavailable: 1
    selector:
      matchExpressions:
      - key: app
        operator: In
        values:
        - rook-ceph-osd
  status:
    conditions:
    - lastTransitionTime: "2025-09-08T22:38:12Z"
      message: ""
      observedGeneration: 5
      reason: SufficientPods
      status: "True"
      type: DisruptionAllowed
    currentHealthy: 6
    desiredHealthy: 5
    disruptionsAllowed: 1
    expectedPods: 6
    observedGeneration: 5
kind: List
metadata:
  resourceVersion: ""
```

elias-dbx commented on pkg/operator/ceph/disruption/clusterdisruption/osd_test.go:
```diff
@@ -427,9 +435,50 @@ func TestReconcilePDBForOSD(t *testing.T) {
+					return zone == tc.expectedDrainingFailureDomainName
+				})
+				for i, zone := range nonDrainingZones {
+					maxUnavailable := intstr.FromInt32(int32(tc.expectedMaxUnavailableCount))
```
ah right, thanks for the tip. fixed.

sp98 commented:
@travisn @elias-dbx I tried node draining test. But didn't see expected results. 
- Drained node `minikube` with OSD (0 and 3)
- Node drain was successful and PGs became unhealthy. 
- Blocking PDBs were added to `minikube-02` and `minikube-03` 

- Node `minikube` was added back. 
- PGs became healthy. 
- blocking PDBs were removed and main PDB was added back. 
- Main PDB still has the match label for OSD-3, even though its running. 

```

2025-09-10 04:21:08.560796 I | clusterdisruption-controller: osd "rook-ceph-osd-0" is down on node "minikube" but no node drain is detected
2025-09-10 04:21:08.560867 I | clusterdisruption-controller: osd "rook-ceph-osd-3" is down on node "minikube" but no node drain is detected
2025-09-10 04:21:09.265936 I | clusterdisruption-controller: OSD(s) [0 3] are down and PGs are not clean. PGs Status: "cluster is not fully clean. PGs: [{StateName:active+undersized+degraded Count:1}]"
2025-09-10 04:21:09.265976 I | clusterdisruption-controller: OSD failure Domains : ["minikube" "minikube-m02" "minikube-m03"]
2025-09-10 04:21:09.265990 I | clusterdisruption-controller: Draining Failure Domain: "minikube"
2025-09-10 04:21:09.266003 I | clusterdisruption-controller: Set noout on draining Failure Domain: "true"
2025-09-10 04:21:09.909642 I | clusterdisruption-controller: reconciling osd pdb controller
2025-09-10 04:21:25.600099 I | clusterdisruption-controller: osd "rook-ceph-osd-3" is down on node "minikube" but no node drain is detected
2025-09-10 04:21:26.511945 I | clusterdisruption-controller: OSD(s) [3] are down but PGs are clean. PG Status: "all PGs in cluster are clean"
2025-09-10 04:21:26.512130 I | clusterdisruption-controller: `maxUnavailable` for the main OSD PDB is set to 1
2025-09-10 04:21:26.512191 I | clusterdisruption-controller: all PGs are active+clean. Restoring default OSD pdb settings
2025-09-10 04:21:26.512317 I | clusterdisruption-controller: creating the default pdb "rook-ceph-osd" with maxUnavailable=1 for all osd
2025-09-10 04:21:26.558127 I | clusterdisruption-controller: deleting temporary blocking pdb with "rook-ceph-osd-host-minikube-m02" with maxUnavailable=0 for "host" failure domain "minikube-m02"
2025-09-10 04:21:26.582294 I | clusterdisruption-controller: deleting temporary blocking pdb with "rook-ceph-osd-host-minikube-m03" with maxUnavailable=0 for "host" failure domain "minikube-m03"
2025-09-10 04:21:28.770031 I | clusterdisruption-controller: successfully reconciled OSD PDB controller


Every 2.0s: kubectl get pdb -n rook-ceph rook-ceph-osd -o yaml                                                                                             dhcp53-151.lab.eng.blr.redhat.com: Wed Sep 10 09:55:16 2025

apiVersion: policy/v1
kind: PodDisruptionBudget
metadata:
  creationTimestamp: "2025-09-10T04:21:26Z"
  generation: 1
  name: rook-ceph-osd
  namespace: rook-ceph
  ownerReferences:
  - apiVersion: ceph.rook.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: CephCluster
    name: rook-ceph
    uid: a3d5b7ec-17bf-499d-9fa7-e8aa7e4bcbac
  resourceVersion: "5390"
  uid: 03f9ef10-10e5-4418-9b21-1a61a2620482
spec:
  maxUnavailable: 1
  selector:
    matchExpressions:
    - key: app
      operator: In
      values:
      - rook-ceph-osd
    - key: osd
      operator: NotIn
      values:
      - "3"
status:
  conditions:
  - lastTransitionTime: "2025-09-10T04:21:26Z"
    message: ""
    observedGeneration: 1
    reason: SufficientPods
    status: "True"
    type: DisruptionAllowed
  currentHealthy: 5
  desiredHealthy: 4
  disruptionsAllowed: 1
  expectedPods: 5
  observedGeneration: 1




❯ kr get pdb
NAME                MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
rook-ceph-mgr-pdb   N/A             1                 1                     16m
rook-ceph-mon-pdb   N/A             1                 1                     16m
rook-ceph-osd       N/A             1                 1                     4m31s



````

elias-dbx commented:
If the PDB is properly created after operator restart that most likely means that we are missing a reconcile trigger in the PDB control loop, like you said.

It looks like this is where we configure the reconcile triggers: https://github.com/rook/rook/blob/281aedcdb615d7872b35e67649415f8f8d77fe61/pkg/operator/ceph/disruption/clusterdisruption/add.go#L108

elias-dbx commented:
Reading the code here, it seems like we most likely have the same bug currently on `master` as well? The PDB only gets reconciled when maxUnavailable=1, so if for example with the current `master` code the PDB has maxUnavilable=3 as some OSDs are down but all PGs are clean it wouldn't get reconciled either, I think.
https://github.com/rook/rook/blob/281aedcdb615d7872b35e67649415f8f8d77fe61/pkg/operator/ceph/disruption/clusterdisruption/add.go#L73
